File: tests/sdk/nodejs/test/cur-iot.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { requireReady, signedJsonTarget, uniquePrefix } from "../lib/helpers.mjs";

test("CUR PutReportDefinition Describe Delete", async (t) => {
  if (!(await requireReady(t))) return;
  const prefix = uniquePrefix();
  const reportName = `sdk-${prefix}`;

  const put = await signedJsonTarget(
    "cur",
    "AWSOrigamiServiceGatewayService.PutReportDefinition",
    {
      ReportDefinition: {
        ReportName: reportName,
        TimeUnit: "MONTHLY",
        Format: "textORcsv",
        Compression: "GZIP",
        S3Bucket: `cur-sdk-${prefix}`,
        S3Prefix: "reports",
        S3Region: "us-east-1",
        AdditionalSchemaElements: ["RESOURCES"],
        ReportVersioning: "OVERWRITE_REPORT",
      },
    },
  );
  assert.equal(put.status, 200, put.body);
  assert.ok(put.json?.ReportName, `missing ReportName: ${put.body}`);

  const desc = await signedJsonTarget(
    "cur",
    "AWSOrigamiServiceGatewayService.DescribeReportDefinitions",
    {},
  );
  assert.equal(desc.status, 200, desc.body);
  assert.ok(
    (desc.json?.ReportDefinitions || []).length >= 1,
    `DescribeReportDefinitions empty: ${desc.body}`,
  );

  const del = await signedJsonTarget(
    "cur",
    "AWSOrigamiServiceGatewayService.DeleteReportDefinition",
    { ReportName: reportName },
  );
  assert.equal(del.status, 200, del.body);
});

test("IoT CreateThing cert AttachThingPrincipal shadow", async (t) => {
  if (!(await requireReady(t))) return;
  const prefix = `sdk-iot-${uniquePrefix()}`;

  const created = await signedJsonTarget("iot", "AWSIotService.CreateThing", {
    thingName: prefix,
    attributePayload: { attributes: { env: "sdk" } },
  });
  assert.equal(created.status, 200, created.body);

  const cert = await signedJsonTarget(
    "iot",
    "AWSIotService.CreateKeysAndCertificate",
    { setAsActive: true },
  );
  assert.equal(cert.status, 200, cert.body);
  const certArn = cert.json?.certificateArn;
  assert.ok(certArn, `missing certificateArn: ${cert.body}`);

  const pol = await signedJsonTarget("iot", "AWSIotService.CreatePolicy", {
    policyName: `${prefix}-pol`,
    policyDocument:
      '{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":"iot:*","Resource":"*"}]}',
  });
  assert.equal(pol.status, 200, pol.body);

  const attach = await signedJsonTarget(
    "iot",
    "AWSIotService.AttachThingPrincipal",
    { thingName: prefix, principal: certArn },
  );
  assert.equal(attach.status, 200, attach.body);

  const shadow = await signedJsonTarget(
    "iot-data",
    "AWSIotDataService.UpdateThingShadow",
    {
      thingName: prefix,
      state: { desired: { color: "green" } },
    },
  );
  assert.equal(shadow.status, 200, shadow.body);

  const got = await signedJsonTarget(
    "iot-data",
    "AWSIotDataService.GetThingShadow",
    { thingName: prefix },
  );
  assert.equal(got.status, 200, got.body);
  assert.ok(got.json?.state, `GetThingShadow missing state: ${got.body}`);
});

File: tests/sdk/go/cur_iot_test.go
package sdk_test

import (
	"testing"
)

func TestCURPutDescribeDelete(t *testing.T) {
	requireReady(t)
	prefix := uniquePrefix(t)

	bucket := "cur-sdk-" + prefix
	// Best-effort: Put succeeds even if bucket missing (status ERROR).
	putStatus, putBody, putParsed := signedJSONTarget(t, "cur", "AWSOrigamiServiceGatewayService.PutReportDefinition", map[string]any{
		"ReportDefinition": map[string]any{
			"ReportName":               "sdk-" + prefix,
			"TimeUnit":                 "MONTHLY",
			"Format":                   "textORcsv",
			"Compression":              "GZIP",
			"S3Bucket":                 bucket,
			"S3Prefix":                 "reports",
			"S3Region":                 "us-east-1",
			"AdditionalSchemaElements": []string{"RESOURCES"},
			"ReportVersioning":         "OVERWRITE_REPORT",
		},
	})
	if putStatus != 200 {
		t.Fatalf("PutReportDefinition status=%d body=%s", putStatus, putBody)
	}
	if name, _ := putParsed["ReportName"].(string); name == "" {
		t.Fatalf("PutReportDefinition missing ReportName: %s", putBody)
	}

	descStatus, descBody, descParsed := signedJSONTarget(t, "cur", "AWSOrigamiServiceGatewayService.DescribeReportDefinitions", map[string]any{})
	if descStatus != 200 {
		t.Fatalf("DescribeReportDefinitions status=%d body=%s", descStatus, descBody)
	}
	defs, _ := descParsed["ReportDefinitions"].([]any)
	if len(defs) < 1 {
		t.Fatalf("DescribeReportDefinitions empty: %s", descBody)
	}

	delStatus, delBody, _ := signedJSONTarget(t, "cur", "AWSOrigamiServiceGatewayService.DeleteReportDefinition", map[string]any{
		"ReportName": "sdk-" + prefix,
	})
	if delStatus != 200 {
		t.Fatalf("DeleteReportDefinition status=%d body=%s", delStatus, delBody)
	}
}

func TestIoTThingCertShadow(t *testing.T) {
	requireReady(t)
	prefix := "sdk-iot-" + uniquePrefix(t)

	createStatus, createBody, _ := signedJSONTarget(t, "iot", "AWSIotService.CreateThing", map[string]any{
		"thingName": prefix,
		"attributePayload": map[string]any{
			"attributes": map[string]string{"env": "sdk"},
		},
	})
	if createStatus != 200 {
		t.Fatalf("CreateThing status=%d body=%s", createStatus, createBody)
	}

	certStatus, certBody, certParsed := signedJSONTarget(t, "iot", "AWSIotService.CreateKeysAndCertificate", map[string]any{
		"setAsActive": true,
	})
	if certStatus != 200 {
		t.Fatalf("CreateKeysAndCertificate status=%d body=%s", certStatus, certBody)
	}
	certARN, _ := certParsed["certificateArn"].(string)
	if certARN == "" {
		t.Fatalf("missing certificateArn: %s", certBody)
	}

	polStatus, polBody, _ := signedJSONTarget(t, "iot", "AWSIotService.CreatePolicy", map[string]any{
		"policyName":     prefix + "-pol",
		"policyDocument": `{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":"iot:*","Resource":"*"}]}`,
	})
	if polStatus != 200 {
		t.Fatalf("CreatePolicy status=%d body=%s", polStatus, polBody)
	}

	attachStatus, attachBody, _ := signedJSONTarget(t, "iot", "AWSIotService.AttachThingPrincipal", map[string]any{
		"thingName": prefix,
		"principal": certARN,
	})
	if attachStatus != 200 {
		t.Fatalf("AttachThingPrincipal status=%d body=%s", attachStatus, attachBody)
	}

	shadowStatus, shadowBody, _ := signedJSONTarget(t, "iot-data", "AWSIotDataService.UpdateThingShadow", map[string]any{
		"thingName": prefix,
		"state": map[string]any{
			"desired": map[string]any{"color": "green"},
		},
	})
	if shadowStatus != 200 {
		t.Fatalf("UpdateThingShadow status=%d body=%s", shadowStatus, shadowBody)
	}

	getStatus, getBody, getParsed := signedJSONTarget(t, "iot-data", "AWSIotDataService.GetThingShadow", map[string]any{
		"thingName": prefix,
	})
	if getStatus != 200 {
		t.Fatalf("GetThingShadow status=%d body=%s", getStatus, getBody)
	}
	state, _ := getParsed["state"].(map[string]any)
	if state == nil {
		t.Fatalf("GetThingShadow missing state: %s", getBody)
	}
}
